api/graphql/reservations/reservation_serializers.py

```diff
@@ -717,34 +721,142 @@ def check_metadata_fields(self, data, reservation_unit) -> None:
+                allowed_values = list(
+                    map(lambda x: x.code, reservation_unit.payment_types.all())
+                )
+                raise ValidationErrorWithCode(
+                    f"Reservation unit does not support {payment_type} payment type. "
+                    f"Allowed values: {', '.join(allowed_values)}",
+                    ValidationErrorCodes.INVALID_PAYMENT_TYPE,
+                )
+        return data
 
     @property
     def validated_data(self):
         validated_data = super().validated_data
 
-        reservation_unit_requires_handling = self.instance.reservation_unit.filter(
-            require_reservation_handling=True
-        ).exists()
-
-        if (
-            reservation_unit_requires_handling
-            or self.instance.applying_for_free_of_charge
-        ):
+        if self._requires_handling():
             validated_data["state"] = STATE_CHOICES.REQUIRES_HANDLING
         else:
             validated_data["state"] = STATE_CHOICES.CONFIRMED
+
         return validated_data
 
     def save(self, **kwargs):
+        self.fields.pop("payment_type")
         instance = super().save(**kwargs)
 
+        state = self.validated_data["state"]
+        if state == STATE_CHOICES.CONFIRMED:
+            payment_type = self.validated_data["payment_type"].upper()
+            price_net = round_decimal(self.instance.price_net, 2)
+            price_vat = round_decimal(
+                self.instance.price_net * (self.instance.tax_percentage_value / 100), 2
+            )
+            price_total = round_decimal(self.instance.price, 2)
+
+            if payment_type == PaymentType.ON_SITE:
+                PaymentOrder.objects.create(
+                    payment_type=payment_type,
+                    status=PaymentStatus.PAID_MANUALLY,
+                    language=self.instance.reservee_language or Language.FI,
+                    price_net=price_net,
+                    price_vat=price_vat,
+                    price_total=price_total,
+                    reservation=self.instance,
+                )
+            else:
+                payment_order = create_verkkokauppa_order(self.instance)
+                PaymentOrder.objects.create(
+                    payment_type=payment_type,
+                    status=PaymentStatus.DRAFT,
+                    language=self.instance.reservee_language or Language.FI,
+                    price_net=price_net,
+                    price_vat=price_vat,
+                    price_total=price_total,
+                    reservation=self.instance,
+                    reservation_user_uuid=self.instance.user.uuid,
+                    order_id=payment_order.order_id,
+                    checkout_url=payment_order.checkout_url,
+                    receipt_url=payment_order.receipt_url,
+                )
+                Reservation.objects.filter(pk=self.instance.pk).update(
```

Comment:
This is the part I wanted to ask you about. The point is to update reservation state AFTER we've created the order.

I moved the `instance = super().save(**kwargs)` up and this whole order mayhen happens after that. If everything goes well, we'll update the state again.

Can you think any better way to do this?

Reply:
I tried to draft the solution in my head, but it felt like it would just open a whole new can of worms 😅  I feel that the main problem is the fact that we don't want this action to be tied to any save but to be specific to confirm action. I guess it would be possible, but the solution would probably be on the same spot on the hackish-solutions scale 😄 

Perhaps I just leave this as is for now and let's improve if we notice that this does not work.

Reply:
Yes, I agree 🤝